app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/VatDetailsController.scala

```diff
@@ -99,4 +125,105 @@ object VatDetailsController {
           submittedAnswer => Seq(Yes, No).contains(submittedAnswer.confirm)))
 
+  val normalizedText: Mapping[String] = of[String].transform(_.replaceAll("\\s", ""), identity)
+
+  val vatDetailsForm: Form[VatDetails] =
+    Form[VatDetails](
+      mapping(
+        "vrn"     -> normalizedText.verifying(validVrn),
+        "regDate" -> validRegDate
+      ) { case (a, b) => VatDetails(Vrn(a), b) } { vatDetails =>
+        Some((vatDetails.vrn.value, vatDetails.regDate))
+      }
+    )
+
+  def nonEmpty(failure: String): Constraint[String] = Constraint[String] { fieldValue: String =>
+    if (fieldValue.trim.isEmpty) Invalid(ValidationError(failure)) else Valid
+  }
+
+  def validVrn: Constraint[String] = Constraint[String] { fieldValue: String =>
+    nonEmpty("vat-details.vrn.required")(fieldValue) match {
+      case i: Invalid => i
+      case Valid =>
+        if (!fieldValue.matches("[0-9]{9}"))
+          Invalid(ValidationError("vat-details.vrn.regex-failure"))
+        else if (!Vrn.isValid(fieldValue.trim.toUpperCase))
+          Invalid(ValidationError("vat-details.vrn.checksum-failure"))
+        else
+          Valid
+    }
+  }
+
+  def validRegDate: Mapping[LocalDate] =
+    tuple(
+      "year"  -> text.verifying("year", y => !y.trim.isEmpty || y.matches("^[0-9]{4}$")),
+      "month" -> text.verifying("month", y => !y.trim.isEmpty || y.matches("^[0-9]{1,2}$")),
+      "day"   -> text.verifying("day", d => !d.trim.isEmpty || d.matches("^[0-9]{1,2}$"))
+    ).verifying(checkOneAtATime(Seq(realDateConstraint, futureDateConstraint)))
+      .transform(
+        { case (y, m, d) => LocalDate.of(y.trim.toInt, m.trim.toInt, d.trim.toInt) },
+        (date: LocalDate) => (date.getYear.toString, date.getMonthValue.toString, date.getDayOfMonth.toString)
+      )
+
+  private def realDateConstraint: Constraint[(String, String, String)] = Constraint[(String, String, String)] {
+    data: (String, String, String) =>
+      val (year, month, day) = data
+      Try {
+        val date = LocalDate.of(year.toInt, month.toInt, day.toInt)
+        if (date.isBefore(LocalDate.of(1900, 1, 1)))
+          Invalid(ValidationError("vat-details.regDate.is.not.real"))
+        else
+          Valid
+      } match {
+        case Failure(_) => Invalid(ValidationError("vat-details.regDate.invalid"))
+        case Success(p) => p
+      }
+  }
+
+  private def futureDateConstraint: Constraint[(String, String, String)] = Constraint[(String, String, String)] {
+    data: (String, String, String) =>
+      val (year, month, day) = data
+      Try {
+        if (LocalDate.of(year.toInt, month.toInt, day.toInt).isAfter(LocalDate.now()))
+          Invalid(ValidationError("vat-details.regDate.must.be.in.past"))
+        else
+          Valid
+      } match {
+        case Failure(_) => Invalid(ValidationError("vat-details.regDate.invalid"))
+        case Success(p) => p
+      }
+  }
+
+  def formWithRefinedErrors(form: Form[VatDetails]): Form[VatDetails] = {
```

Comment:
all the below validation is a bit of duplication from DateOfBirth validation , i will refactor this in the upcoming PR